### ISSUE TITLE
Freeway motorway link destinations are not displayed before entering the junction

### ISSUE THREAD

**Issue by dead10ck** [osmandapp/Osmand#8767]
I've tried this in quite a few places now, and it seems like osmAnd does not display the name of freeway exits before you enter the junction. Turn destinations are displayed on regular streets, and the name of freeway entrances are displayed before entering the freeway onramp (assuming things are tagged correctly, of course), but when exiting the freeway, osmAnd will not display the name of the exit. It's important to know the name of the exit one has to take, so this is making navigation difficult. Here are some screenshots (with osmAnd 3.6.3 from F-Droid) of a route that goes from 849VCW6F+J9 to 849VCW75+Q3. It should display something to the effect of "exit Old Middlefield Way", but instead it just says to get in the right lane and take a slight right.

![Screenshot_20200407-081814](https://user-images.githubusercontent.com/846275/78747728-f3885780-791e-11ea-8874-644d1ca4f435.jpg)

I've looked at the map data for motorway link way, and it appears everything is tagged correctly. You can even see that osmAnd knows the name of the destination by looking at the detailed instructions:

![Screenshot_20200407-082846](https://user-images.githubusercontent.com/846275/78747779-174b9d80-791f-11ea-8769-3e88bfe394cb.jpg)

but it simply does not display it during navigation.

Another example is going from 849VFV39+MX to 849VFR8V+53. It should say "Exit Willow Road", but it does not.

![Screenshot_20200407-085004](https://user-images.githubusercontent.com/846275/78747816-33e7d580-791f-11ea-8f2c-6e2778c591f6.jpg)

The exit is tagged correctly, and the name appears in the detailed instructions:

![Screenshot_20200407-085250](https://user-images.githubusercontent.com/846275/78747866-54179480-791f-11ea-8806-4f7074e8aa9a.jpg)

I'm using osmAnd 3.6.3 from F-Droid on a OnePlus 7T on Oxygen OS version 10.0.8.HD65AA.

I'm using the California San Francisco offline map, updated with osmAnd Live, and the screenshots were taken with the last map change recorded at 2020-04-07 08:18 PDT.

**Comment by sonora** [comment]
I think the issue is that e.g. https://www.openstreetmap.org/node/65381289 does not have a name. If you tag that with what the ground-truth road sign displays, it should work just fine.

**Comment by dead10ck** [comment]
@sonora according to [the wiki](https://wiki.openstreetmap.org/wiki/Tag:highway%3Dmotorway_junction), the name tag on a motorway junction is for the name of the junction itself, not for the destination, and most junctions in the US do not have names, just numbers.

>`name=*` for the name of the junction or interchange. *Do not confuse the **name of a junction** with the **destination(s) the junction leads to**. In most cases worldwide, sign information describes destinations, not the name of the junction or interchange itself.* If a signpost or indication displays destinations exclusively, this data belongs to `destination=*` tags, **not** the `name=*` of the `highway=motorway_junction` node.

So the navigator needs to get the info from the `destination*` tags on the `motorway_link` way, not the `name` tag on the `motorway_junction` node.

**Comment by sonora** [comment]
@dead10ck , I had checked 2 of your cases (101/Old Middlefield Way, and 101/Willow Road), and did for both cases get the correct pre-announcement, about 1 mile out, saying "_In about 1 mile, keep right onto exit 399B_". This worked for both my California map of August 2017 as well as the current San Fran map downloaded earlier today.

And once you are on the ramp (i.e., more specifically, after you passed the motor_junction node), the following turn announcements correctly contain "_... towards Old Middlefield Way_", later "_... onto Old Middlefield Way_".

If you do  not get any of these announcements, we must be checking for an issue somewhere in your settings or setup:

- Maybe first check your tickmarks in `Car profile / Configure Profile / Navigation Settings / Voice prompts`,  specifically make sure "`Street names (TTS)`" is selected.

- If that is the case, please temporarily enable the "Development plugin" (under `Menu button / Plugins`), then in any profile go to `Settings / Plugin Settings / Test voice prompts`, and let me know what you hear when you tap e.g. test buttons 3.7, 3.11, and 3.12.

Regarding a motorway:_junction's "Name" field: True, those are common in some countries, uncommon in others. If I remember our code correctly, if the motor_junction has no name but has a "destination" field filled, that should also correctly be announced.

I am quite certain the issue you see here (at least the portion I can reproduce of announcing just exit numbers, and destinations only after the exit node is passed) comes from the fact that the exit to take (being the next node on your route) in all your cases only has only a ref number, hence no other information to announce at that point.

**Comment by sonora** [comment]
PS: Quick screenshot from AFTER you passed the motorway_junction node:

![Screenshot_20200408-172547_OsmAnd~_](https://user-images.githubusercontent.com/845735/78802851-95ab4c80-79be-11ea-9110-edd0a3d77367.jpg)

**Comment by dead10ck** [comment]
> If you do not get any of these announcements, we must be checking for an issue somewhere in your settings or setup:
> 
> * Maybe first check your tickmarks in `Car profile / Configure Profile / Navigation Settings / Voice prompts`,  specifically make sure "`Street names (TTS)`" is selected.
> 
> * If that is the case, please temporarily enable the "Development plugin" (under `Menu button / Plugins`), then in any profile go to `Settings / Plugin Settings / Test voice prompts`, and let me know what you hear when you tap e.g. test buttons 3.7, 3.11, and 3.12.
> 

I tried this, and I do not hear any announcements either. I enabled the developer plugin and went into the "Test voice prompts" section, and clicking buttons does nothing (I made sure my volume was up).

However, voice prompts are not the problem for me—I actually do not want voice prompts, just an announcement on the screen.

> @dead10ck , I had checked 2 of your cases (101/Old Middlefield Way, and 101/Willow Road), and did for both cases get the correct pre-announcement, about 1 mile out, saying "_In about 1 mile, keep right onto exit 399B_". This worked for both my California map of August 2017 as well as the current San Fran map downloaded earlier today.
> 
> And once you are on the ramp (i.e., more specifically, after you passed the motor_junction node), the following turn announcements correctly contain "_... towards Old Middlefield Way_", later "_... onto Old Middlefield Way_".

I actually do get the announcement for "turn on Old Middlefield Way", just like in your screenshot, but as I said earlier, getting this announcement *after* I am already on the motorway link is far too late to be helpful.

However, you say that you get an announcement for "next exit at 399B"—as you can see from the screenshots, I do not see anything on the screen that even says an exit number. Do you see anything on your screen, or do you only hear a voice prompt?

> Regarding a motorway:_junction's "Name" field: True, those are common in some countries, uncommon in others. If I remember our code correctly, if the motor_junction has no name but has a "destination" field filled, that should also correctly be announced.
> 
> I am quite certain the issue you see here (at least the portion I can reproduce of announcing just exit numbers, and destinations only after the exit node is passed) comes from the fact that the exit to take (being the next node on your route) in all your cases only has only a ref number, hence no other information to announce at that point.

That is because the `destination=*` tag is not on the `motorway_junction` node, but on the `motorway_link` way that immediately follows the `motorway_junction` node. According to the wiki for how these things should be mapped and tagged, it needs to get the announcement information from the junction node *and* the `motorway_link` way that branches from the junction node.

**Comment by rickmastfan67** [comment]
@dead10ck I see the Exit number listed for the US-101 one.  "Middlefield Way 399".  However, I do agree that there should maybe be shown differently, maybe like "Middlefield Way, Exit #399" instead.

**Comment by dead10ck** [comment]
@rickmastfan67 interesting, could you post a screenshot? If it's displaying the destination and ref number for you before you get to the junction node, then maybe this is some error to my particular device, or perhaps the F-Droid build.

**Comment by rickmastfan67** [comment]
I was just referring to your pictures above.

**Comment by dead10ck** [comment]
@rickmastfan67 are you talking about in the detailed step-by-step instructions? The problem I'm talking about is that that information doesn't appear on the navigation screen. I can't pull up the detailed instructions while I'm driving to see what the exit street or number is. It should be displayed at the top of the screen, just as any turn on a normal street would.

**Comment by sonora** [comment]
3 takeaways here from this issue, I summarize as follows:

(1) If your voice prompts do not work, we can debug this, please open a separate issue. This lies in the regime of hardware settings, Android TTS config, and audio output channel. Tapping audio prompt test button 11.1 would give me decisive hints.

(2) The January implementation of announcing exits takes the related information from the respective motor_junction's `ref`, `name`, and `destination` tags. I think this is the correct way of doing it. (Certainly debatable, but 'pre-reading' destination tags of subsequent road segments seems questionable, we do not do this elsewhere either.)

(3) While voice prompting works fine for all motor_junctions, it seems that the on-screen display fails to pop up for exits just tagged with a ref (and neither a name nor a destination.) I have re-opened #1347 to investigate and fix.

**Comment by vshcherb** [comment]
@sonora please open for (3) a separate issue, unfortunately I will follow up on it in 3.7

**Comment by sonora** [comment]
I have looked at the latest code and it should be fine. I have also re-tested the Old Middlefield Way situation, see new screenshot. All should be good!  ;)

![Screenshot_20200410-192302_OsmAnd~](https://user-images.githubusercontent.com/845735/79009916-fc686b80-7b60-11ea-8cb7-066a8989e6f7.jpg)

**Comment by dead10ck** [comment]
Awesome! I eagerly await the next release! Thanks for testing.

**Comment by dead10ck** [comment]
@vshcherb @sonora sorry, I really don't mean to be pushy or naggy, but I'm just noticing it seems the normal release cadence is on the order of months. Is there any way the fix for the missing freeway exit notification could be released as a point release on its own? It is unfortunately a major usability issue.

**Comment by sonora** [comment]
For testing you could use the nightly builds. As far as minor releases are concerned, this is for Victor to decide.

**Comment by vshcherb** [comment]
(y)

**Comment by Neuwessi11** [comment]
Sorry for re-opening, but this issue seems to be similar to my problem.

On every German autobahn the name of the exit (which is tagged on the junction node) is written on a sign 1000 m ahead the junction. This is very well shown in OsmAnd.
But ALL signs I see later (on the street) don't show the name of the junction, but show the cities that I have to tag AFTER the junction node. Because these are the cities I can see on the important signs before the exit, it is necessary to see them on my display, but OsmAnd never announces them. 
And because it is necessary that  in a navigation app I find the signs that I see in nature, I must contradict @sonora:

You wrote on 8 Apr 2020:
> (2) The January implementation of announcing exits takes the related information from the respective motor_junction's `ref`, `name`, and `destination` tags. I think this is the correct way of doing it. (Certainly debatable, but 'pre-reading' destination tags of subsequent road segments seems questionable, we do not do this elsewhere either.)
> 
> EDIT: Upon re-reading the OSM wiki, maybe you are right, it seems to indicate that the `destination=*` tag should be placed '_on the way way that branches off from the node motorway_junction_'

I cannot see that OsmAnd takes the relevant information from the 'destination' tags. And if so, it provides this information far too late, because I need this information before entering the exit, but if it is shown (what is normally not) it is shown after passing the junction node, when it is completely useless.

Even more strange in Sweden.
Driving there you see (in nature) a small yellow sign with 'trafikplats xxx', which is to be tagged as the name of the junction. And this is shown by OsmAnd.
But driving the 1000 m to the exit you NEVER will see this junction name again. Instead of that you get signs with cities, that in some cases have really nothing to do with that trafikplats-sign. 
But OsmAnd shows and shows and shows the trafikplats-sign until I pass the junction node, what is really useless.
